File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        answer = [0] * len(temperatures)
        stack = []
        
        for i, curr in enumerate(temperatures):
            while stack and curr > temperatures[stack[-1]]:
                prev = stack.pop()
                answer[prev] = i - prev
            stack.append(i)
        
        return answer

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> stack;  // (index, temperature)
        vector<int> res(temperatures.size(), 0);  
        
        for (int i = 0; i < temperatures.size(); i++) {
            int t = temperatures[i];
            while (!stack.empty() && stack.top().second < t) {
                auto pair = stack.top();
                stack.pop();
                res[pair.first] = i - pair.first;
            }
            stack.push({i, t});
        }
        return res;
    }
};